File: src-tauri/src/dto/backup_restore.rs
//! Data Transfer Objects for backup and restore operations (Slice 12).

use serde::{Deserialize, Serialize};

/// Result of a successful backup export.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupResult {
    /// Absolute path where the backup was written.
    pub path: String,
    /// Number of bytes written.
    pub bytes: u64,
    /// Schema version at time of backup (last applied migration version).
    pub schema_version: i64,
}

/// Input for the restore command.
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreInput {
    /// Absolute path to the backup file to restore.
    pub backup_path: String,
    /// User must explicitly confirm this is a destructive operation.
    pub confirmed: bool,
}

/// Validation result for a backup file before restoring.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreValidation {
    /// Whether the file appears to be a valid SQLite database.
    pub is_valid_sqlite: bool,
    /// Whether all expected Caduxo tables are present.
    pub has_expected_schema: bool,
    /// Whether the schema version is compatible (same as or lower than current).
    pub is_version_compatible: bool,
    /// Human-readable description of each validation check (English canonical
    /// fallback text). Kept as the source of truth for older frontends.
    pub checks: Vec<String>,
    /// Stable check codes aligned with `checks`. Each entry is `Some(code)`
    /// when the service emits a typed code for that check, or `None` when
    /// the row has no localized counterpart (reserved for future additions).
    /// The frontend dispatches by code and falls back to the matching
    /// `checks[i]` string when the code is `None` or unrecognized.
    pub check_codes: Vec<Option<String>>,
    /// The schema version detected in the backup, if readable.
    pub detected_schema_version: Option<i64>,
    /// Whether the backup passed all checks and is safe to restore.
    pub can_restore: bool,
}

/// Result of a successful restore operation.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreResult {
    /// The path to the restored database file.
    pub database_path: String,
    /// Schema version after restore.
    pub schema_version: i64,
}

File: src/lib/backup_restore.ts
/**
 * TypeScript API wrapper for backup and restore commands (Slice 12).
 */

import { invoke } from "@tauri-apps/api/core";
import { save, open } from "@tauri-apps/plugin-dialog";

// ─── DTOs (mirror Rust DTOs in src-tauri/src/dto/backup_restore.rs) ────────────

/** Result of a successful backup export. */
export interface BackupResult {
 path: string;
 bytes: number;
 schemaVersion: number;
}

/** Input for the restore command. */
export interface RestoreInput {
 backupPath: string;
 confirmed: boolean;
}

/** Validation result for a backup file before restoring. */
export interface RestoreValidation {
 isValidSqlite: boolean;
 hasExpectedSchema: boolean;
 isVersionCompatible: boolean;
 /** Human-readable English fallback text for each check. */
 checks: string[];
 /**
  * Stable check codes aligned by index with `checks`. Each entry is `Some`
  * (a string) when the service emits a typed code for that check, or `null`
  * when the row has no localized counterpart. The frontend dispatches by
  * code and falls back to the matching `checks[i]` string when the code is
  * `null` / `undefined` or unrecognized.
  */
 checkCodes: (string | null)[];
 detectedSchemaVersion: number | null;
 canRestore: boolean;
}

/** Result of a successful restore operation. */
export interface RestoreResult {
 databasePath: string;
 schemaVersion: number;
}

// ─── Command wrappers ─────────────────────────────────────────────────────────

/**
 * Exports a copy of the active database to a user-chosen destination.
 * Uses a native save dialog to let the user pick the destination.
 */
export async function exportBackupWithDialog(): Promise<BackupResult | null> {
 const path = await save({
  defaultPath: `caduxo-backup-${new Date().toISOString().slice(0, 10)}.db`,
  filters: [{ name: "SQLite Database", extensions: ["db"] }],
  title: "Export database backup",
 });

 if (!path) {
  return null; // User cancelled
 }

 return invoke<BackupResult>("export_backup", { destination: path });
}

/**
 * Validates a backup file before restoring.
 * Opens a native file picker so the user can select a backup file.
 */
export async function validateBackupWithDialog(): Promise<{
 path: string;
 validation: RestoreValidation;
} | null> {
 const path = await open({
  multiple: false,
  filters: [
   { name: "SQLite Database", extensions: ["db", "sqlite", "sqlite3"] },
  ],
  title: "Select backup file to validate",
 });

 if (!path || typeof path !== "string") {
  return null; // User cancelled or multiple files returned
 }

 const validation = await invoke<RestoreValidation>("validate_backup", {
  backupPath: path,
 });

 return { path, validation };
}

/**
 * Restores the database from a validated backup file.
 * Requires explicit user confirmation (the frontend is responsible for
 * displaying a confirmation dialog).
 */
export async function restoreBackup(
 input: RestoreInput,
): Promise<RestoreResult> {
 return invoke<RestoreResult>("restore_backup", { input });
}
